hoomd/md/Thermostat.h

```diff
@@ -337,38 +337,128 @@ class MTTKThermostat : public Thermostat
         {
         m_thermo->compute(timestep);
 
-        const auto ntdof = m_thermo->getTranslationalDOF();
-        const auto nrdof = m_thermo->getRotationalDOF();
+        Scalar e_factor;
+        Scalar rt_left;
+        Scalar rr_left;
+        const int ntdof = static_cast<int>(m_thermo->getTranslationalDOF());
+        const int nrdof = static_cast<int>(m_thermo->getRotationalDOF());
+        const auto ket_int = m_thermo->getTranslationalKineticEnergy();
+        const auto ker_int = m_thermo->getRotationalKineticEnergy();
         if ((ntdof != 0 && m_thermo->getTranslationalKineticEnergy() == 0)
             || (nrdof != 0 && m_thermo->getRotationalKineticEnergy() == 0))
             {
             throw std::runtime_error("Bussi thermostat requires non-zero initial temperatures");
             }
+
         unsigned int instance_id = 0;
         if (m_group->getNumMembersGlobal() > 0)
             instance_id = m_group->getMemberTag(0);
         RandomGenerator rng(Seed(RNGIdentifier::BussiThermostat, timestep, m_sysdef->getSeed()),
                             instance_id);
 
         const auto set_T = m_T->operator()(timestep);
-        GammaDistribution<double> gamma_translation(ntdof / 2.0, set_T);
-        GammaDistribution<double> gamma_rotation(nrdof / 2.0, set_T);
 
-        return {std::sqrt(gamma_translation(rng) / m_thermo->getTranslationalKineticEnergy()),
-                std::sqrt(gamma_rotation(rng) / m_thermo->getRotationalKineticEnergy())};
+        if (deltaT < 1e-12)
+            {
+            e_factor = 1.0;
+            }
+        else if (m_tau / deltaT > 0.1)
+            {
+            e_factor = exp(-deltaT / m_tau);
+            }
+        else // the limit case when tau is near 0
+            {
+            e_factor = 0.0;
+            }
+
+        NormalDistribution<double> normal_translation(1.0);
+        NormalDistribution<double> normal_rotation(1.0);
+        Scalar rt = normal_translation(rng);
+        Scalar rr = normal_rotation(rng);
+
+        const int ntdof_left = ntdof - 1;
+        const int nrdof_left = nrdof - 1;
+        const bool ntdof_left_even = ntdof_left % 2 == 0;
+        const bool nrdof_left_even = nrdof_left % 2 == 0;
+        if (ntdof_left == 0)
+            {
+            rt_left = 0.;
+            }
+        else if (ntdof_left == 1)
+            {
+            Scalar rt_temp = normal_translation(rng);
+            rt_left = rt_temp * rt_temp;
+            }
+        else if (ntdof_left_even)
+            {
+            GammaDistribution<double> gamma_translation(ntdof_left / 2.0, 1.0);
+            rt_left = 2.0 * gamma_translation(rng);
+            }
+        else
+            {
+            GammaDistribution<double> gamma_translation((ntdof_left - 1) / 2.0, 1.0);
+            Scalar rt_temp = normal_translation(rng);
+            rt_left = 2.0 * gamma_translation(rng) + rt_temp * rt_temp;
+            }
```

Comment:
It is not clear what this sequence of conditions achieves. You mentioned it came from a correction to the original article. Is that correction published? We need to cite an explanation for this logic for future reference.

Reply:
@joaander The correct implementation written by Bussi himself is here:
https://github.com/GiovanniBussi/StochasticVelocityRescaling/blob/master/resamplekin.cpp

Specifically, the `resamplekin_sumnoises` function.

Reply:
Thanks for confirming this, Martin. I suspected as such. Alex, please simplify the code accordingly.